source/core/shape.cc

```diff
@@ -875,19 +894,53 @@ void
 RBFShape<dim>::update_precalculations(DoFHandler<dim> &dof_handler,
                                       std::shared_ptr<Mapping<dim>> /*mapping*/)
 {
-  int maximal_level = dof_handler.get_triangulation().n_levels();
+  const int maximal_level = dof_handler.get_triangulation().n_levels();
 
   for (int level = highest_level_searched + 1; level < maximal_level; level++)
     {
       const auto &cell_iterator = dof_handler.cell_iterators_on_level(level);
       for (const auto &cell : cell_iterator)
         {
+          // We first check if we are in the hierarchy levels where we simply
+          // assume that children have the same likely nodes as their parents.
+          if (level > maximal_level - 1 - levels_not_precalculated)
+            {
+              likely_nodes_map[cell] = likely_nodes_map[cell->parent()];
+              continue;
+            }
+          // If the precalculations have reached the finest level, then the cell
+          // precalculation only has to be done for cells that are locally
+          // owned.
           if (level == maximal_level)
             if (!cell->is_locally_owned())
-              break;
-          determine_likely_nodes_for_one_cell(cell, cell->vertex(0));
+              continue;
+          // In the same way, the cell precalculation only has to be done on
+          // cells that aren't artificial (in other words, locally owned or
+          // ghost cells)
+          if (!cell->is_artificial_on_level())
+            determine_likely_nodes_for_one_cell(cell, cell->barycenter());
         }
       highest_level_searched = level;
```

Comment:
You need a way to clear all the variable you seems to cache when the shape change position

Reply:
I'm adding a function for that